lib/dialects/abstract/query-generator/query-proto.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+const { Composition } = require('./composition');
+
+/*
+ * Named collections of Composition elements.
+ *
+ * Useful for organizing and manipulating the different parts of a query.
+ * Allow disordered query building and easy substitution of query parts. Flexible,
+ * non-nested and non-exahustive organization.
+ */
+
+class QueryProto {
+  constructor(proto) {
+    if (proto) {
+      this.set(proto);
+    } else {
+      // Should clone Composition, NOT copy shallow references
+      for (const key of this.constructor.partNames) {
+        this[key] = new Composition();
+      }
+    }
+  }
+  static get partNames() {
+    return [];
+  }
+  set(proto) {
+    // Should clone proto into this object, NOT copy shallow references
+
+    for (const key of this.constructor.partNames) {
+      this[key] = new Composition();
+
+      if (proto[key] && proto[key].length) {
+        this[key] = proto[key].clone();
+      } else {
+        this[key] = new Composition();
+      }
+    }
+
+    return this;
+  }
+  clone() {
+    // Should clone Composition, NOT copy shallow references
+    const clone = new this.constructor();
+
+    for (const key of this.constructor.partNames) {
+      clone[key] = this[key].clone();
+    }
+
+    return clone;
+  }
+  toComposition() {
+    throw new Error(`Class ${this.constructor.name} has not implemented toComposition method`);
+  }
+}
+
+class SelectProto extends QueryProto {
+  static get partNames() {
+    return Object.freeze(['attributes', 'from', 'join', 'where', 'group',
```

Comment:
Not what I meant, my bad.

I was trying to suggest create the array outside of the class and pass a reference instead.

Reply:
I did that originally, but in a previous review I was told to use `static get fn() {}` instead

Reply:
Or perhaps you mean to define the array outside?
```
const selectProtoPartNames = Object.freeze([...]);

class SelectProto extends QueryProto {
  static get partNames() {
    return selectProtoPartNames;
  }
}
```

I thought that JS engines were allowed to cache static getters, but it can also be done this way.

Reply:
I don't think getters can be cached at all, since it needs to return dynamic values, I'd just do `SelectProto.partNames = Object.freeze(...)`